pkg/network/go/bininspect/pclntab.go

```diff
@@ -0,0 +1,284 @@
+		}
+		p.funcData = sectionAccess{
+			section:    p.section,
+			baseOffset: int64(offset(7)),
+		}
+		p.funcTable = sectionAccess{
+			section:    p.section,
+			baseOffset: int64(offset(7)),
+		}
+	case ver116:
+		p.funcTableSize = uint32(offset(0))
+		p.funcNameTable = sectionAccess{
+			section:    p.section,
+			baseOffset: int64(offset(2)),
+		}
+		p.funcData = sectionAccess{
+			section:    p.section,
+			baseOffset: int64(offset(6)),
+		}
+		p.funcTable = sectionAccess{
+			section:    p.section,
+			baseOffset: int64(offset(6)),
+		}
+	}
+
+	return nil
+}
+
+// uintptr returns the pointer-sized value encoded at b.
+// The pointer size is dictated by the table being read.
+// based on https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L186.
+func (p *pclntanSymbolParser) uintptr(b []byte) uint64 {
+	if p.ptrSize == 4 {
+		return uint64(p.byteOrderParser.Uint32(b))
+	}
+	return p.byteOrderParser.Uint64(b)
+}
+
+// getFuncTableFieldSize returns the size of a field in the function table.
+// based on https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L388-L392
+func getFuncTableFieldSize(version version, ptrSize int) int {
+	if version >= ver118 {
+		return 4
+	}
+	return ptrSize
+}
+
+// getSymbols returns the symbols from the pclntab section that match the symbol filter.
+// based on https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L300-L329
+func (p *pclntanSymbolParser) getSymbols() (map[string]*elf.Symbol, error) {
+	symbols := make(map[string]*elf.Symbol, p.symbolFilter.getNumWanted())
+	data := sectionAccess{section: p.section}
+	for currentIdx := uint32(0); currentIdx < p.funcTableSize; currentIdx++ {
+		// based on https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L315
+		_, err := p.funcTable.ReadAt(p.funcTableBuffer, int64((2*currentIdx+1)*uint32(p.funcTableFieldSize)))
+		if err != nil {
+			continue
+		}
+
+		// based on https://github.com/golang/go/blob/6a861010be9eed02d5285509cbaf3fb26d2c5041/src/debug/gosym/pclntab.go#L321
+		data.baseOffset = int64(p.uint(p.funcTableBuffer)) + p.funcData.baseOffset
+		nameOffset := field(p.ptrSize, p.cachedVersion, p.byteOrderParser, data, p.ptrBufferSizeHelper)
+		funcName := p.funcName(nameOffset)
+
+		if funcName == "" {
+			continue
+		}
+		symbols[funcName] = &elf.Symbol{
+			Name: funcName,
+		}
+		if len(symbols) == p.symbolFilter.getNumWanted() {
+			break
+		}
+	}
+	if len(symbols) < p.symbolFilter.getNumWanted() {
+		return symbols, ErrFailedToFindAllSymbols
+	}
+	return symbols, nil
+}
+
+// funcName returns the name of the function found at off.
+func (p *pclntanSymbolParser) funcName(off uint32) string {
+	n, err := p.funcNameTable.ReadAt(p.funcNameHelper, int64(off))
+	if err != nil {
+		return ""
+	}
+	if p.symbolFilter.want(string(p.funcNameHelper[:n])) {
```

Comment:
This string can have NULs, right? Guess it's not an issue for the filtering but a comment could be useful.

Reply:
handled in the code